Sources/Utils/Backtrace.cpp

```diff
@@ -22,20 +22,54 @@
 #include <sstream>
 #endif
 
+#if defined(ANDROID)
+// Android provides libunwind in the NDK instead of glibc backtrace
+#include <unwind.h>
+#endif
+
 namespace ds2 {
 namespace Utils {
 
-#if defined(OS_DARWIN) || (defined(__GLIBC__) && !defined(PLATFORM_TIZEN))
+#if defined(ANDROID)
+struct unwind_state {
+    void **buffer_cursor;
+    void **buffer_end;
+};
+
+static _Unwind_Reason_Code UnwindTrace(struct _Unwind_Context* context,
+                                       void* arg) {
+  struct unwind_state* state = reinterpret_cast<struct unwind_state*>(arg);
+  uintptr_t ip = _Unwind_GetIP(context);
+  if (ip != 0) {
+    if (state->buffer_cursor >= state->buffer_end)
+      return _URC_END_OF_STACK; // Backtrace will be truncated
+
+    *(state->buffer_cursor++) = reinterpret_cast<void*>(ip);
+  }
+  return _URC_NO_REASON;
+}
+
+// An approximation of glibc's backtrace using _Unwind_Backtrace
+static int backtrace(void *stack_buffer[], int stack_size) {
+  unwind_state state = {
+    .buffer_cursor = stack_buffer,
+    .buffer_end = stack_buffer + stack_size,
+  };
+
+  _Unwind_Backtrace(UnwindTrace, &state);
+  return state.buffer_cursor - stack_buffer;
+}
+#endif
+
+#if defined(ANDROID) || defined(OS_DARWIN) || (defined(__GLIBC__) && !defined(PLATFORM_TIZEN))
 static void PrintBacktraceEntrySimple(void *address) {
   DS2LOG(Error, "%" PRI_PTR, PRI_PTR_CAST(address));
 }
-#endif
 
-#if defined(OS_DARWIN) || (defined(__GLIBC__) && !defined(PLATFORM_TIZEN))
 void PrintBacktrace() {
   static const int kStackSize = 100;
   static void *stack[kStackSize];
-  int stackEntries = ::backtrace(stack, kStackSize);
+  int stackEntries = backtrace(stack, kStackSize);
```

Comment:
Should we provide an overload here that will just forward to `::backtrace` for readability?

Reply:
Sure. I went back-and-forth on that when I was making the change. Overriding `backtrace` was a bit less code so I chose that.